---
title: "Transcriptome analysis"
subtitle: "anya project"
params:
  genome: !!r here::here("genome", "stutzeri_genome.fasta")
  seqs: !!r here::here("seqs", "seqs.tsv")
  auxilliary: !!r here::here("genome", "additional_seqs.tsv")
  gff: !!r here::here("genome", "annotation.gff")
  column_data: !!r here::here("seqs", "column_data.tsv")
  go: !!r here::here("genome", "go_annot.csv")
  pathway: !!r here::here("genome", "pathways.csv")
---

#### Markdown option
```{r}
knitr::opts_chunk$set(echo=FALSE, message = FALSE)
```

#### Load packages
```{r message=FALSE, echo=FALSE}
source("utilityFuns.R")
library(tidyverse)
library(RUVSeq)
library(janitor)
```

#### Calculate number of cores
```{r}
# Detect available cores
core_number <- parallel::detectCores()-4
# Initialize cluster
cl <- parallel::makeCluster(core_number)
```

#### Load annotation files
```{r}
# Load gff file
gtfRegionList <- clean_gff(annotation_file=params$gff)

# Load additional samples metadata
column_data <- clean_column_metadata(params$column_data)
```

#### Mapping
```{r}
proj <- func_align(genome = params$genome,
                   seqs_file = params$seqs,
                   aux = params$auxilliary)
```


#### Check mappings
```{r}
stats <- alignmentStats(proj)
```

#### Quantification
```{r}
result <- qCount(proj, gtfRegionList, orientation = "same")
```

#### Render quantification and gene product tibbles
```{r}
# Create counts tibble
result_df <- make_tibble(gtfRegionList, result)$counts_df
```

#### Prepare data from EDA and normalization
```{r}
set <- prep_eda(column_data = column_data,
                counts_matrix = result)
```

#### Choose index for unwanted variance (k, RUVs)
```{r}
pick_k(column_data = column_data,
       eda_set = set)
```

#### Rectify EDA set matrix, based on k=2 from previous step
```{r}
differences <- makeGroups(column_data$group)
set_s <- RUVs(set, unique(rownames(set)), k=2, differences)
```

#### Compare initial set vs rectified set_s objects
#### Compare RLE plots
```{r}
compare_RLE(column_data = column_data,
            initial_set = set,
            rectified_set = set_s)
```

#### Compare PCA plots
```{r}
compare_PCA(column_data = column_data,
            initial_set = set,
            rectified_set = set_s)
```

#### Render heatmap
```{r}
heat_map <- render_heatmap(column_data = column_data,
                           rectified_set = set_s)
```

#### Re-run DESeq2 with the computed covariates (padj < 0.05)
```{r}
re_run_deseq <- re_run_deseq(count_data = subset(result, select=c(-width)),
                   column_data = column_data,
                   rectified_set = set_s,
                   design=as.formula("~group"))

PHE_result <- re_run_deseq$PHE
res_PHE <- re_run_deseq$res_PHE
PY_result <- re_run_deseq$PY
res_PY <- re_run_deseq$res_PY
```

#### GO and KEGG Annotation
```{r message=FALSE}
PHE_pathway <- bio_function(go_file = params$go,
                            pathway_file = params$pathway,
                            log_file = PHE_result)
PHE_up <- PHE_pathway$UP
PHE_down <- PHE_pathway$DOWN
PHE_both <- PHE_pathway$BOTH
  
PY_pathway <- bio_function(go_file = params$go,
                            pathway_file = params$pathway,
                            log_file = PY_result)
PY_up <- PY_pathway$UP
PY_down <- PY_pathway$DOWN
PY_both <- PY_pathway$BOTH
```

### Rename columns from EN to RO
```{r}
```




#### Get PHE and PY metadata
```{r}
PHE_kegg_up <- PHE_up %>% count(pathway_name, sort=T, name="up")
PHE_kegg_down <- PHE_down %>% count(pathway_name, sort=T, name="down")

PY_kegg_up <- PY_up %>% count(pathway_name, sort=T, name="up")
PY_kegg_down <- PY_down %>% count(pathway_name, sort=T, name="down")
```

#### Render final heatmap
```{r}
hm <- render_final_hp(rectified_set = set_s,
                      column_data = column_data,
                      row_data = BOTH)
```

#### Save heatplot and tables
```{r}
# Save heatplots
ggplot2::ggsave(plot=hm, "heatmap.eps", device="eps", path="analysis_results",
               width=11, height = 10)
ggplot2::ggsave(plot=hm, "heatmap.pdf", device="pdf", path="analysis_results",
               width=11, height = 10)

# Save log2FoldChange
readr::write_csv(PY_up, "analysis_results/py_log_up.csv")
readr::write_csv(PY_down, "analysis_results/py_log_down.csv")
readr::write_csv(PY_both, "analysis_results/py_log_both.csv")
readr::write_csv(PY_kegg_up, "analysis_results/py_kegg_up.csv")
readr::write_csv(PY_kegg_down, "analysis_results/py_kegg_down.csv")
readr::write_csv(PHE_up, "analysis_results/phe_log_up.csv")
readr::write_csv(PHE_down, "analysis_results/phe_log_down.csv")
readr::write_csv(PHE_both, "analysis_results/phe_log_both.csv")
readr::write_csv(PHE_kegg_up, "analysis_results/phe_kegg_up.csv")
readr::write_csv(PHE_kegg_down, "analysis_results/phe_kegg_down.csv")
```


```{r}
PY_UP_2 <- readr::read_csv("analysis_results/py_log_up.csv") %>% 
  select(gene, pathway_name)
PY_DOWN_2 <- readr::read_csv("analysis_results/py_log_down.csv") %>% 
  select(gene, pathway_name)

PHE_UP_2 <-  readr::read_csv("analysis_results/phe_log_up.csv") %>% 
  select(gene, pathway_name)
PHE_DOWN_2 <- readr::read_csv("analysis_results/phe_log_down.csv") %>% 
  select(gene, pathway_name)

BOTH_2 <- bind_rows(PY_UP_2, PY_DOWN_2, PHE_UP_2, PHE_DOWN_2) %>% 
  distinct(gene, .keep_all = T) %>% 
  rename(`CÄƒi metabolice` = pathway_name,
         Gene = gene)

column_data_2 <- column_data %>% 
  mutate(Treatment = case_when(str_detect(Treatment, "phenanthrene") ~ "fenantren", 
                               str_detect(Treatment, "pyrene") ~ "piren",
                               TRUE ~ Treatment)) %>% 
  rename(Tratament = Treatment)

hm_2 <- render_final_hp_RO(rectified_set = set_s,
                      column_data = column_data_2,
                      row_data = BOTH_2)



PY_UP_waffle <- get_waffle(PY_UP_2)
ggsave("analysis_results/PY_UP_waffle.tiff", plot = PY_UP_waffle)
PY_DOWN_waffle <- get_waffle(PY_DOWN_2)
ggsave("analysis_results/PY_DOWN_waffle.tiff", plot = PY_DOWN_waffle)

PHE_UP_waffle <- get_waffle(PHE_UP_2)
ggsave("analysis_results/PHE_UP_waffle.tiff", plot = PHE_UP_waffle)
PHE_DOWN_waffle <- get_waffle(PHE_DOWN_2)
ggsave("analysis_results/PHE_DOWN_waffle.tiff", plot = PHE_DOWN_waffle)




```







